testutil/fuse.go

```diff
@@ -0,0 +1,250 @@
+	return h.fs.Unlink(name, context)
+}
+
+func (h *HookFs) GetXAttr(name string, attribute string, context *fuse.Context) ([]byte, fuse.Status) {
+	return h.fs.GetXAttr(name, attribute, context)
+}
+
+func (h *HookFs) ListXAttr(name string, context *fuse.Context) ([]string, fuse.Status) {
+	return h.fs.ListXAttr(name, context)
+}
+
+func (h *HookFs) RemoveXAttr(name string, attr string, context *fuse.Context) fuse.Status {
+	return h.fs.RemoveXAttr(name, attr, context)
+}
+
+func (h *HookFs) SetXAttr(name string, attr string, data []byte, flags int, context *fuse.Context) fuse.Status {
+	return h.fs.SetXAttr(name, attr, data, flags, context)
+}
+
+func (h *HookFs) OnMount(nodeFs *pathfs.PathNodeFs) {
+	h.fs.OnMount(nodeFs)
+}
+
+func (h *HookFs) OnUnmount() {
+	h.fs.OnUnmount()
+}
+
+func (h *HookFs) Open(name string, flags uint32, context *fuse.Context) (nodefs.File, fuse.Status) {
+	return h.fs.Open(name, flags, context)
+}
+
+func (h *HookFs) Create(name string, flags uint32, mode uint32, context *fuse.Context) (nodefs.File, fuse.Status) {
+	return h.fs.Create(name, flags, mode, context)
+}
+
+func (h *HookFs) OpenDir(name string, context *fuse.Context) ([]fuse.DirEntry, fuse.Status) {
+	return h.fs.OpenDir(name, context)
+}
+
+func (h *HookFs) Symlink(value string, linkName string, context *fuse.Context) fuse.Status {
+	return h.fs.Symlink(value, linkName, context)
+}
+
+func (h *HookFs) Readlink(name string, context *fuse.Context) (string, fuse.Status) {
+	return h.fs.Readlink(name, context)
+}
+
+func (h *HookFs) StatFs(name string) *fuse.StatfsOut {
+	return h.fs.StatFs(name)
+}
+
+func (h *HookFs) NewServe() (*fuse.Server, error) {
+	server, err := newHookServer(h)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+//tests will want to run this in a  goroutine.
+func (h *HookFs) Start(server *fuse.Server) {
+	server.Serve()
+}
+
+func newHookServer(hookfs *HookFs) (*fuse.Server, error) {
+	opts := &nodefs.Options{
+		NegativeTimeout: time.Second,
+		AttrTimeout:     time.Second,
+		EntryTimeout:    time.Second,
+	}
+	pathFsOpts := &pathfs.PathNodeFsOptions{ClientInodes: true}
+	pathFs := pathfs.NewPathNodeFs(hookfs, pathFsOpts)
+	conn := nodefs.NewFileSystemConnector(pathFs.Root(), opts)
+	originalAbs, _ := filepath.Abs(hookfs.Original)
+	mOpts := &fuse.MountOptions{
+		AllowOther: true,
+		Name:       hookfs.FsName,
+		FsName:     originalAbs,
+	}
+	server, err := fuse.NewServer(conn.RawFS(), hookfs.Mountpoint, mOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+func NewFuseServer(t *testing.T, original, mountpoint string, hook Hook) (interface{}, error) {
+	createDirIfAbsent(original)
+	createDirIfAbsent(mountpoint)
+	fs, err := NewHookFs(original, mountpoint, hook)
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := fs.NewServe()
+	if err != nil {
+		return nil, err
+	}
+
+	//async start fuse server
+	go func() {
+		fs.Start(server)
```

Comment:
where is this stopped ?

Reply:
> where is this stopped ?

If call `syscall.Unmount`, it will stop. And it will be called in CleanUp method

Reply:
thanks, could you put a comment to clarify this.

Reply:
Done